ipa-core/src/seq_join.rs

```diff
@@ -95,209 +113,348 @@ pub trait SeqJoin {
+                    // Making futures cancellable is critical to avoid hangs.
+                    // if one of them panics, unwinding causes spawner to drop and, in turn,
+                    // it blocks the thread to await all pending futures completion. If there is
+                    // a dependency between futures, pending one will never complete.
+                    // Cancellable futures will be cancelled when spawner is dropped which is
+                    // the behavior we want.
+                    this.spawner
+                        .spawn_cancellable(f.into_future().instrument(Span::current()), || {
+                            panic!("cancelled")
+                        });
+                } else {
+                    break;
+                }
+            }
+
+            // Poll spawner if it has work to do. If both source and spawner are empty, we're done
+            if this.spawner.remaining() > 0 {
+                this.spawner.as_mut().poll_next(cx).map(|v| match v {
+                    Some(Ok(v)) => Some(v),
+                    Some(Err(_)) => panic!("task is cancelled"),
+                    None => None,
+                })
+            } else if this.source.is_done() {
+                Poll::Ready(None)
+            } else {
+                Poll::Pending
+            }
         }
 
-        let active = NonZeroUsize::new(10).unwrap();
-        let res = seq_try_join_all(active, (1..5).map(f)).await.unwrap();
-        assert_eq!((1..5).collect::<Vec<_>>(), res);
+        fn size_hint(&self) -> (usize, Option<usize>) {
+            let in_progress = self.spawner.remaining();
+            let (lower, upper) = self.source.size_hint();
+            (
+                lower.saturating_add(in_progress),
+                upper.and_then(|u| u.checked_add(in_progress)),
+            )
+        }
     }
 
-    #[tokio::test]
-    async fn try_join_early_abort() {
-        const ERROR: &str = "error message";
-        fn f(i: u32) -> impl Future<Output = Result<u32, &'static str>> {
-            lazy(move |_| match i {
-                1 => Ok(1),
-                2 => Err(ERROR),
-                _ => panic!("should have aborted earlier"),
-            })
-        }
+    /// TODO: change it to impl Future once https://github.com/rust-lang/rust/pull/115822 is
+    /// available in stable Rust.
+    pub(super) fn parallel_join<'fut, I, F, O, E>(iterable: I) -> BoxFuture<'fut, Result<Vec<O>, E>>
+    where
+        I: IntoIterator<Item = F> + Send,
+        F: Future<Output = Result<O, E>> + Send + 'fut,
+        O: Send + 'static,
+        E: Send + 'static,
+    {
+        // TODO: implement spawner for shuttle
+        let mut scope = {
+            let iter = iterable.into_iter();
+            // SAFETY: scope object does not escape this function. All futures are driven to
+            // completion inside it or cancelled if a panic occurs.
+            let mut scope = unsafe { create_spawner() };
```

Comment:
The async stuff makes it harder to reason about "escapes this function". If you view the function as being active until the boxed future completes, then yes, I think that's true. But if anyone calls `mem::forget` for the boxed future before it completes, then there could be a memory safety violation.

It is my understanding though, that as long as you don't `mem::forget` the future (and thus instead always either drive it to completion or drop it), then there can't be a memory safety violation. Maybe it would make sense to put `mem::forget` on the disallowed function list like `join_all`?

I also thought about whether we could use the safe `spawn_and_block` API. That would mean consuming a thread per active query, which seems tolerable. But if we want to avoid calling `spawn_and_block` from a regular tokio executor, and instead call it via `thread::spawn` or `tokio::task::spawn_blocking`, that brings us right back to the `'static` problem. The only way I can see it working is if we spawn some number of tasks that receive futures over channels in `query::executor::do_query` (which is where the query lifetime is realized -- it is already calling `tokio::spawn`).

Reply:
Or maybe `thread::scope` helps us here?

Reply:
You are right - I mistakenly assumed that not leaking futures created within `Scope` makes it safe. It does not and I wrote a test to confirm that. 

> It is my understanding though, that as long as you don't mem::forget the future (and thus instead always either drive it to completion or drop it), then there can't be a memory safety violation. Maybe it would make sense to put mem::forget on the disallowed function list like join_all?

that's my plan - restricting `mem::forget` won't be enough, we also need to monitor usages of `Box::leak` and `ManuallyDrop`. I wouldn't go down that path if we were a library, but for IPA as an application this could be acceptable mitigation along with sanitizers run. 

> I also thought about whether we could use the safe spawn_and_block API. 

I am going to play more with it but I don't feel that we can make it work. We combine the use of `seq_join` with `parallel_join` inside it and blocking the thread may lead to some weird deadlocks. My current thinking that it is probably not worth trying to make it safe, given that `seq_join` is inherently unsafe and cannot be make safe.